### PR TITLE
[Validator] check for __get method existence if property is uninitialized

### LINKED ISSUE #35544: Validator incompatibility with ProxyManager
**Symfony version(s) affected**: 5.0.4

**Description**  
When trying to validate a value holder proxy generated with ocramius/proxy-manager, all the public properties will be unset at construction.
This results in `ReflectionProperty::isInitialized()` returning false as clearly documented [here](https://www.php.net/manual/en/reflectionproperty.isinitialized.php).
PR #35532 introduced check on `isInitialized` returning null instead of firing an `__get` call.

**Possible Solution**  
Check if `__get` method exists before blindly returning null, as per rfc "reads from uninitialized properties will generate a TypeError unless __get() is defined".

### PR DESCRIPTION
**Pull request by alekitto**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #35544
| License       | MIT

Resolve bug #35544.

On PHP 7.4, check if object implements `__get` magic method if property is reported as uninitialized before returning null.

### CODE REVIEW

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (xabbuh) [inline src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: Does this mean that there was a bug in PHP 7.4.0 and `__get()` should never have been called? In this case, I am not convinced that we should do anything here as the behaviour is inline with what happens in PHP 7.4.1+.

Comment (xabbuh) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: see also https://3v4l.org/8E83P

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: or should we call `__get()` directly?

Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: There is a discussion [PHP Bug # 78904] (https://bugs.php.net/bug.php?id=78904).

This is a PHP related problem, I think. Framework should not implement any hack for this. ProxyManager and others are currently not compatible with PHP 7.4.1+.

Comment (alekitto) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: `__get` **must** be called when accessing a property, if the property has been explicitly unset, but not if it is uninitialized.
On the other side `ReflectionProperty::isInitialized` is returning false for both situations and there is no *safe* way to distinguish one state from the other.

Comment (alekitto) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: @greedyivan The problem is not limited to the ProxyManager and I don't think that symfony could stop supporting PHP 7.4.1+ until the discussion on PHP is resolved.

Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: Here is the case:
```php
class Foo
{
    public string $unintialized;
    public string $unsetted;
    
    public function __construct() {
        unset($this->unsetted);
    }

    public function __get($var)
      {
        return 'not set';
     }
}

$foo = new Foo();

assert(null === $foo->unintialized) // validator should return null here
assert('not set' === $foo->unsetted); // validator should return the actual value
```

Comment (alekitto) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: @greedyivan It is exactly what this patch does

**Review by xabbuh — CHANGES_REQUESTED** [review]
I am strongly against merging this as is as this would now work differently than how PHP behaves in 7.4.1+. I suggest to wait until a final decision has been made for PHP and then decide whether or not we have to adapt the validator code.

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: @alekitto, agree. I've just check it out.

Only two thinks:
- `$reflProperty->hasType()` means nothing here
- `} catch (\TypeError $e) {` - here must be `\Error`

**Comment by greedyivan** [comment]
> @xabbuh this would now work differently than how PHP behaves in 7.4.1+.

Here we trying to solve one special case, when typed property is `unset` in constructor. In that state `__get` method is invoked when the property is accessed.

Only difference with 7.4.0 that this magic method was invoked in uninitialized state also.

This is normal (invoking `__get` for unset properties)  but not clearly documented feature. ProxyManager uses it, and previous bugfix broke this special case.

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (alekitto) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: The `hasType` has been added as if the property is not typed, no additional check should be made.
I think that `\Error` is too generic. Per typed properties RFC `\TypeError` is raised in case of access an uninitialized non-nullable typed property.

Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: > The `hasType` has been added as if the property is not typed, no additional check should be made.
We have to do some check:
- for non-typed in this case we check `hasType`
- for typed we check `hasType` and `isInitialized`
Using only one check `isInitialized` does the same and has no additional cost.

> I think that `\Error` is too generic. Per typed properties RFC `\TypeError` is raised in case of access an uninitialized non-nullable typed property.
https://3v4l.org/lmTYH

Don't trust docs blindly.

Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: > https://3v4l.org/lmTYH
According that, additional tests should be added.

- typed uninitialized propery with no unset in constructor and with __get method.

This test would validate a correct way of `\Error` catching. In case if it will be changed to `\TypeError` in future.

Comment (alekitto) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: I've just added a check. Too bad that RFC is not respected about the `\TypeError`

Comment (greedyivan) [reply src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: > Too bad that RFC is not respected about the `\TypeError`

[Doc Bug #79206](https://bugs.php.net/bug.php?id=79206)
Let's see, it is a doc bug or not.

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: of *if* it is

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: this comment should be moved inside the catch block below and be updated: it's not "TypeError" anymore

File: src/Symfony/Component/Validator/Mapping/PropertyMetadata.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Mapping/PropertyMetadata.php]: we shouldn't mention any versions of PHP in the comment
we should make the code work consistently in all versions as much as possible.

**Review by nicolas-grekas — COMMENTED** [review]
Reasonable, isn't it?

**Comment by alekitto** [comment]
> Reasonable, isn't it?

Yes, i've corrected the comments.

**Review by nicolas-grekas — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @alekitto.

**Pull request merged**
